modules/tensor_mechanics/include/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.h

```diff
@@ -0,0 +1,226 @@
+   * can then be used in the inheriting classes to calculate cross slip interactions.
+   * The values of number_cross_slip_directions and number_cross_slip_planes must be set to
+   * use this sorting method.
+   */
+  void sortCrossSlipFamilies();
+
+  /**
+   * A helper method for inherting classes to identify to which cross slip family vector
+   * a particular slip system index belongs after the slip systems have been sorted. Returns
+   * the integer value of the identified cross slip system family for the outer vector
+   * created in sortCrossSlipFamilies.
+   */
+  unsigned int identifyCrossSlipFamily(const unsigned int index);
+
+  /**
+   * This virtual method is called to set the constitutive internal state variables
+   * current value and the previous substep value to the old property value for
+   * the start of the stress convergence while loop.
+   */
+  virtual void setInitialConstitutiveVariableValues() {}
+
+  /**
+   * This virtual method is called to set the current constitutive internal state
+   * variable value to that of the previous substep at the beginning of the next
+   * substep increment. In cases where only one substep is taken (or when the first)
+   * substep is taken, this method sets the current constitutive internal state
+   * variable value to the old value.
+   */
+  virtual void setSubstepConstitutiveVariableValues() {}
+
+  /**
+   * Stores the current value of the constitutive internal state variables into
+   * a separate material property in case substepping is required, once the
+   * constitutive variables have passed convergence tolerances. This separate
+   * material property is used as the previous substep value in the associated
+   * setSubstepConstitutiveVariableValues method in the next substep (if taken).
+   */
+  virtual void updateSubstepConstitutiveVariableValues() {}
+
+  /**
+   * This virtual method is called to calculate the slip system slip
+   * increment based on the constitutive model defined in the child class.
+   * This method must be overwritten in the child class.
+   */
+  virtual bool calculateSlipRate() = 0;
+
+  void calculateEquivalentSlipIncrement(RankTwoTensor & /*equivalent_slip_increment*/);
+
+  /*
+   * This virtual method is called to find the derivative of the slip increment
+   * with respect to the applied shear stress on the slip system based on the
+   * constiutive model defined in the child class.  This method must be overwritten
+   * in the child class.
+   */
+  virtual void calculateConstitutiveSlipDerivative(std::vector<Real> & /*dslip_dtau*/) = 0;
+
+  /*
+   * Finalizes the values of the state variables and slip system resistance
+   * for the current timestep after convergence has been reached.
+   */
+
+  virtual void cacheStateVariablesBeforeUpdate() {}
+
+  virtual void calculateStateVariableEvolutionRateComponent() {}
+
+  virtual bool updateStateVariable() = 0;
+
+  virtual void calculateSlipResistance() {}
+
+  /*
```

Comment:
make sure you have a double star on opening the comment for doxygen to pick it up

Reply:
Oh, got it. I did not know the double star is for doxygen.